### PR TITLE
Data Ingestion Improvement/Cleanup/Bug Fix - Part 2

### LINKED ISSUE #221: Handle big text ingestion
Recently, I saw below error and feel that we need a mechanism to handle large docs
```
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:439} WARNING - Error during upsert. Rolling back all inserts for docs with errors.
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:328} INFO - Removing id f9c660b6-a393-54ef-9f95-20a616c3edfe for rollback.
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:333} INFO - UUID 767b916a-f6ef-51e1-bf23-cea43a801baa does not exist. Skipping deletion during rollback.
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:328} INFO - Removing id c7ccb038-c35a-501e-8404-1ce26782f0f7 for rollback.
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:328} INFO - Removing id 8a084e49-5e06-5827-8165-69c05d29716e for rollback.
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:328} INFO - Removing id 9a413399-6b68-5660-99d8-4bda030f9496 for rollback.
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:467} ERROR - Errors encountered during ingest.
[2023-12-19, 05:42:30 UTC] {ask_astro_weaviate_hook.py:468} ERROR - <generator object AskAstroWeaviateHook.ingest_data.<locals>.<genexpr> at 0x7fc00df27e00>
[2023-12-19, 05:42:31 UTC] {taskinstance.py:1937} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/airflow/decorators/base.py", line 221, in execute
    return_value = super().execute(context)
                   ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/airflow/operators/python.py", line 192, in execute
    return_value = self.execute_callable()
                   ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/airflow/operators/python.py", line 209, in execute_callable
    return self.python_callable(*self.op_args, **self.op_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/airflow/include/tasks/extract/utils/weaviate/ask_astro_weaviate_hook.py", line 469, in ingest_data
    raise AirflowException("Errors encountered during ingest.")
airflow.exceptions.AirflowException: Errors encountered during ingest.
```
https://clmkpsyfc000301kcjie00t1l.astronomer.run/d2uulcqu/dags/ask_astro_load_airflow_docs/grid?dag_run_id=scheduled__2023-12-18T05%3A00%3A00%2B00%3A00&task_id=ingest_data&tab=mapped_tasks&map_index=0

### LINKED ISSUE #258: Research & Implement: Data Ingestion Related Improvements
### Context
- This is a follow-up action item sub-issue as a result of this research issue https://github.com/astronomer/ask-astro/issues/133
- The items mentioned below will be investigated to see if improvements can be made with their various implementations. If so, make the related PRs to implement the changes.

### Items of Investigation
1. **Data Cleaning and Exclusion of Irrelevant Content**
   - Implement data cleaning during ingestion to remove non-essential content like navigation bars, footers, headers, and other irrelevant sections that may introduce keyword spam and reduce retrieval accuracy.

2. **Review and Refinement of Chunking Logic**
   - Reassess the logic for document chunking to prevent the inclusion of headers or short, meaningless text segments.

3. **Summarization of Large Documents**
   - Generate and insert summaries for excessively large documents that are split into numerous chunks, using a language model to aid in comprehension and retrieval.

4. **Topic Keyword Extraction and Metadata Storage**
   - Perform topic keyword extraction on each document in the Vector DB, store the results as metadata, and enhance queries with user-prompt-derived keywords during Q&A sessions. This strategy requires significant effort and its effectiveness is yet to be determined.

### LINKED ISSUE #285: LangSmith Log Anomaly: ValidationError none is not an allowed value
**Describe the bug**
Saw recurrence of the error ValidationError none is not an allowed value after 0.2.0 release.

Example Log
```
ValidationError(model='Document', errors=[{'loc': ('page_content',), 'msg': 'none is not an allowed value', 'type': 'type_error.none.not_allowed'}])Traceback (most recent call last):


  File "/app/.venv/lib/python3.11/site-packages/langchain/chains/base.py", line 306, in __call__
    self._call(inputs, run_manager=run_manager)


  File "/app/.venv/lib/python3.11/site-packages/langchain/chains/conversational_retrieval/base.py", line 155, in _call
    docs = self._get_docs(new_question, inputs, run_manager=_run_manager)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain/chains/conversational_retrieval/base.py", line 317, in _get_docs
    docs = self.retriever.get_relevant_documents(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 211, in get_relevant_documents
    raise e


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 204, in get_relevant_documents
    result = self._get_relevant_documents(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain/retrievers/contextual_compression.py", line 44, in _get_relevant_documents
    docs = self.base_retriever.get_relevant_documents(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 211, in get_relevant_documents
    raise e


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 204, in get_relevant_documents
    result = self._get_relevant_documents(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain/retrievers/contextual_compression.py", line 44, in _get_relevant_documents
    docs = self.base_retriever.get_relevant_documents(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 211, in get_relevant_documents
    raise e


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 204, in get_relevant_documents
    result = self._get_relevant_documents(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain/retrievers/multi_query.py", line 175, in _get_relevant_documents
    documents = self.retrieve_documents(queries, run_manager)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain/retrievers/multi_query.py", line 210, in retrieve_documents
    docs = self.retriever.get_relevant_documents(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 211, in get_relevant_documents
    raise e


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/retrievers.py", line 204, in get_relevant_documents
    result = self._get_relevant_documents(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain_community/retrievers/weaviate_hybrid_search.py", line 161, in _get_relevant_documents
    docs.append(Document(page_content=text, metadata=res))
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


  File "/app/.venv/lib/python3.11/site-packages/langchain_core/load/serializable.py", line 107, in __init__
    super().__init__(**kwargs)


  File "/app/.venv/lib/python3.11/site-packages/pydantic/v1/main.py", line 341, in __init__
    raise validation_error


pydantic.v1.error_wrappers.ValidationError: 1 validation error for Document
page_content
  none is not an allowed value (type=type_error.none.not_allowed)
  ```

**To Reproduce**
Examples in LangSmith: https://smith.langchain.com/o/4942ae8b-e5be-4964-bb1f-7544886fa06d/projects/p/421b4053-56ab-4704-a77d-33c374e4eac1?columnVisibilityModel=%7B%22outputs%22%3Afalse%2C%22feedback_stats%22%3Atrue%2C%22reference_example%22%3Afalse%2C%22first_token_time%22%3Afalse%2C%22tags%22%3Afalse%2C%22metadata%22%3Afalse%7D&timeModel=%7B%22duration%22%3A%227d%22%7D&peek=accb37fa-6dc6-46f7-bf3c-79daf88e2d97

**Expected behavior**

Should not error out

### LINKED ISSUE #295: Cost Reduction for Ask Astro
## Main Issues
1. Reduce reranker from returning 10 documents to 8 documents
2. clean up data processing steps during ingestion to enforce smaller document length during request

### Migrate to using OpenAI's GPT-3.5 models as they are newer and cheaper by 5X. 

**OpenAI**
The `0125` model is the same as the 16k model in Azure
image

**Azure OpenAI**
image

### Migrate to new OpenAI's embedding model (5X cheaper embedding cost)

From OpenAI: "
#### Reduced price
text-embedding-3-small is also substantially more efficient than our previous generation text-embedding-ada-002 model. 
Pricing for text-embedding-3-small has therefore been reduced by 5X compared to text-embedding-ada-002, from a price per 1k tokens of $0.0001 to $0.00002.
"

Already tracked by https://github.com/astronomer/ask-astro/issues/286

### PR DESCRIPTION
**Pull request by davidgxue**
Code ready for review, evaluations/testing is still in progress

### Description
- This is **part 2** of the data ingestion improvements. The goal is to analyze and investigate any potential issues with the data ingestion process, remove noisy data such as badly formatted content or gibberish text, find bugs, and enforce standards on the process to have predictable cost estimation/reduction.

### Technical Changes
- Renamed the original `split.py` file to `chunking_utils.py` (multiple files changed due to import naming)
- During the above change, found 2 DAGs somehow were NOT changed. Provider docs and SDK docs's incremental ingest DAGs somehow never had a chunking/splitting task (even though the bulk ingest does). I added chunking to these 2 DAGs
- Airflow Docs extraction process has been improved(see code for details)
- Provider Docs extraction process has been improved (see code for details)
- Astro SDK Docs extraction process has been improved (see code for details) -> Astro SDK docs are not currently being ingested, but fixing code since they exist.
- Provider docs used to ingest all past versions leading to many duplicate docs in the DB. It is now changes to only ingest the `stable` version of the docs.

### Evaluations
- Answer quality generally improved with some add on of more concise answering and linking hyperlinks when the model knows the answer is not exhaustive.
- Please attached for a quick list of results
- 
[data_ingest_comparison_part_2.csv](https://github.com/astronomer/ask-astro/files/14484682/data_ingest_comparison_part_2.csv)
[data_ingest_results_part_2.csv](https://github.com/astronomer/ask-astro/files/14484683/data_ingest_results_part_2.csv)
- Bulk ingest also runs with no issues
image

### Related Issues
closes #221
closes #258 
closes #295 (Reranker has been addressed in GCP environment variables, embedding model change completed in a different PR)
closes #285 (This PR prevents empty docs from being ingested)

### CODE REVIEW

File: airflow/include/tasks/extract/astronomer_providers_docs.py [hunk 1]
```diff
@@ -1,14 +1,47 @@
...
+    # df = urls_to_dataframe(urls, "astronomer-providers")
+
+    docs_url_parts = urllib3.parse.urlsplit(base_url)
+    docs_url_base = f"{docs_url_parts.scheme}://{docs_url_parts.netloc}"
+
+    # make sure we didn't accidentally pickup any unrelated links in recursion
+    # get rid of older versions of the docs, only care about "stable" version docs
+    old_version_doc_pattern = r"/(\d+\.)*\d+/"
+    non_doc_links = {
+        link if (docs_url_base not in link) or re.search(old_version_doc_pattern, link) or "/latest/" in link else ""
```
Comment (Lee-W) [inline airflow/include/tasks/extract/astronomer_providers_docs.py:27]: nitpick: should we make `(docs_url_base not in link) or re.search(old_version_doc_pattern, link) or "/latest/"` a function for better readability?

**Review by Lee-W — APPROVED** [review]
Left 1 nitpick. Rest looks good to me

File: airflow/include/tasks/extract/astronomer_providers_docs.py [hunk 2]
```diff
@@ -1,14 +1,47 @@
...
 
 
 def extract_provider_docs() -> list[pd.DataFrame]:
     exclude_docs = ["_api", "_modules", "_sources", "changelog.html", "genindex.html", "py-modindex.html", "#"]
     base_url = "https://astronomer-providers.readthedocs.io/en/stable/"
 
-    urls = get_internal_links(base_url, exclude_docs)
-    df = urls_to_dataframe(urls, "astronomer-providers")
+    all_links = get_internal_links(base_url, exclude_docs)
+    # df = urls_to_dataframe(urls, "astronomer-providers")
```
Comment (pankajastro) [inline airflow/include/tasks/extract/astronomer_providers_docs.py:18]:
```suggestion
```
nit: if not used please remove it

**Review by pankajastro — APPROVED** [review]
LGTM. But this might require a fresh ingestion for testing

**Review by sunank200 — COMMENTED** [review]
@davidgxue requesting you to do the fresh ingestion to the new Database and test the changes before merging.

**Comment by davidgxue** [comment]
Updated the PR description with test results!!

**Pull request merged**